File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/user/UserDataSource.java
package com.heaton.funnyvote.data.user;

import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.retrofit.Server;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.Callback;

public interface UserDataSource {
    User getUser();

    void setUser(User user);

    void removeUser();

    Observable<String> getGuestUserCode(String name);

    void getUserInfo(Callback<Server.UserDataQuery> callback, User user);

    Observable<Server.UserDataQuery> getUserInfo(User user);

    Observable<User> getUser(boolean forceUpdateUserCode);

    void setGuestName(String guestName);

    Observable<?> registerUser(final String appId, final User user, final boolean mergeGuest);

    void unregisterUser();

    Observable<String> getUserCode(String userType, String appId, User user);

    void linkGuestToLoginUser(String otp, String guest, Callback<ResponseBody> callback);

    Observable<ResponseBody> linkGuestToLoginUser(String otp, String guest);

    void changeUserName(Callback<ResponseBody> callback, String tokenType, String token, String name);

    Observable<ResponseBody> changeUserName(String tokenType, String token, String name);

    Observable<?> changeCurrentUserName(final String name);
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/user/UserDataSource.kt
package com.heaton.funnyvote.data.user

import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.retrofit.Server

import okhttp3.ResponseBody
import retrofit2.Callback

/**
 * Created by chiu_mac on 2016/12/6.
 */

interface UserDataSource {

    var user: User

    interface GetUserCallback {
        fun onResponse(user: User)

        fun onFailure()
    }

    interface GetUserInfoCallback {
        fun onResponse(userData: Server.UserDataQuery)

        fun onFailure()
    }

    interface RegisterUserCallback {
        fun onSuccess()

        fun onFailure()
    }

    interface ChangeUserNameCallback {
        fun onSuccess()

        fun onFailure()
    }

    interface GetUserCodeCallback {
        fun onSuccess(userCode: String)

        fun onFalure()
    }

    fun removeUser()

    fun getGuestUserCode(callback: GetUserCodeCallback, name: String)

    fun getUserInfo(callback: Callback<Server.UserDataQuery>, user: User)

    fun getUser(callback: GetUserCallback, forceUpdateUserCode: Boolean)

    fun setGuestName(guestName: String)

    fun registerUser(appId: String, user: User, mergeGuest: Boolean, callback: RegisterUserCallback)

    fun unregisterUser()

    fun getUserCode(userType: String, appId: String, user: User, callback: GetUserCodeCallback)

    fun linkGuestToLoginUser(otp: String, guest: String, callback: Callback<ResponseBody>)

    fun changeUserName(callback: Callback<ResponseBody>, tokenType: String, token: String, name: String)

    fun changeCurrentUserName(name: String, callback: ChangeUserNameCallback)

}
